lib/trsort.c

```diff
@@ -460,90 +460,583 @@ tr_introsort(saidx_t *ISA, const saidx_t *ISAd,
+            STACK_PUSH(ISAd, b, last, limit);
+            STACK_PUSH(ISAd, first, a, limit);
             ISAd += incr, first = a, last = b, limit = next;
           }
         } else {
           if((a - first) <= (b - a)) {
             if(1 < (last - b)) {
-              STACK_PUSH5(ISAd + incr, a, b, next, trlink);
-              STACK_PUSH5(ISAd, first, a, limit, trlink);
+              STACK_PUSH(ISAd + incr, a, b, next);
+              STACK_PUSH(ISAd, first, a, limit);
               first = b;
             } else if(1 < (a - first)) {
-              STACK_PUSH5(ISAd + incr, a, b, next, trlink);
+              STACK_PUSH(ISAd + incr, a, b, next);
               last = a;
             } else {
               ISAd += incr, first = a, last = b, limit = next;
             }
           } else if((last - b) <= (b - a)) {
             if(1 < (last - b)) {
-              STACK_PUSH5(ISAd, first, a, limit, trlink);
-              STACK_PUSH5(ISAd + incr, a, b, next, trlink);
+              STACK_PUSH(ISAd, first, a, limit);
+              STACK_PUSH(ISAd + incr, a, b, next);
               first = b;
             } else {
-              STACK_PUSH5(ISAd, first, a, limit, trlink);
+              STACK_PUSH(ISAd, first, a, limit);
               ISAd += incr, first = a, last = b, limit = next;
             }
           } else {
-            STACK_PUSH5(ISAd, first, a, limit, trlink);
-            STACK_PUSH5(ISAd, b, last, limit, trlink);
+            STACK_PUSH(ISAd, first, a, limit);
+            STACK_PUSH(ISAd, b, last, limit);
             ISAd += incr, first = a, last = b, limit = next;
           }
         }
       } else {
-        if((1 < (b - a)) && (0 <= trlink)) { stack[trlink].d = -1; }
+        if(1 < (b - a)) { partial_copy_flag = 1; }
         if((a - first) <= (last - b)) {
           if(1 < (a - first)) {
-            STACK_PUSH5(ISAd, b, last, limit, trlink);
+            STACK_PUSH(ISAd, b, last, limit);
             last = a;
           } else if(1 < (last - b)) {
             first = b;
           } else {
-            STACK_POP5(ISAd, first, last, limit, trlink);
+            STACK_POP(ISAd, first, last, limit);
           }
         } else {
           if(1 < (last - b)) {
-            STACK_PUSH5(ISAd, first, a, limit, trlink);
+            STACK_PUSH(ISAd, first, a, limit);
             first = b;
           } else if(1 < (a - first)) {
             last = a;
           } else {
-            STACK_POP5(ISAd, first, last, limit, trlink);
+            STACK_POP(ISAd, first, last, limit);
           }
         }
       }
     } else {
       if(trbudget_check(budget, last - first)) {
         limit = tr_ilg(last - first), ISAd += incr;
       } else {
-        if(0 <= trlink) { stack[trlink].d = -1; }
-        STACK_POP5(ISAd, first, last, limit, trlink);
+        partial_copy_flag = 1;
+        STACK_POP(ISAd, first, last, limit);
       }
     }
   }
 #undef STACK_SIZE
 }
 
+static
```

Comment:
I can't tell straigth from the code if it's equal in terms of computation but it seems by boost he refers to "improve" and not to the software library. The code looks like a modification of yuta mori's code and integrates quite closely. It would be nice to get an explaination for trsort and the changes as it's rather unoblivious and almost no documentation exists.